src/linux/init/LSWInit.cpp

```diff
@@ -0,0 +1,689 @@
+        response.Errno = errno;
+        return;
+    }
+    else if (result == 0) // Timed out
+    {
+        response.State = LSWProcessStateRunning;
+        response.Errno = 0;
+        return;
+    }
+
+    if (WI_IsFlagSet(pollResult.revents, POLLIN))
+    {
+        siginfo_t childState{};
+        auto result = waitid(P_PIDFD, process.get(), &childState, WEXITED);
+        if (result < 0)
+        {
+            LOG_ERROR("waitid({}) failed, {}", process.get(), errno);
+            response.Errno = errno;
+            return;
+        }
+
+        response.Code = childState.si_status;
+        response.Errno = 0;
+        response.State = childState.si_code == CLD_EXITED ? LSWProcessStateExited : LSWProcessStateSignaled;
+        return;
+    }
+
+    LOG_ERROR("Poll returned an unexpected error state on fd: {} for pid: ", process.get(), Message.Pid);
+}
+
+void HandleMessageImpl(wsl::shared::SocketChannel& Channel, const LSW_SIGNAL& Message, const gsl::span<gsl::byte>& Buffer)
+{
+    auto result = kill(Message.Pid, Message.Signal);
+    Channel.SendResultMessage(result < 0 ? errno : 0);
+}
+
+template <typename TMessage, typename... Args>
+void HandleMessage(wsl::shared::SocketChannel& Channel, LX_MESSAGE_TYPE Type, const gsl::span<gsl::byte>& Buffer)
+{
+    if (TMessage::Type == Type)
+    {
+        if (Buffer.size() < sizeof(TMessage))
+        {
+            LOG_ERROR("Received message {}, but size is too small: {}. Expected {}", Type, Buffer.size(), sizeof(TMessage));
+            THROW_ERRNO(EINVAL);
+        }
+
+        const auto Message = gslhelpers::try_get_struct<TMessage>(Buffer);
+        HandleMessageImpl(Channel, *Message, Buffer);
+
+        return;
+    }
+    else
+    {
+        if constexpr (sizeof...(Args) > 0)
+        {
+            HandleMessage<Args...>(Channel, Type, Buffer);
+        }
+        else
+        {
+            LOG_ERROR("Received unknown message type: {}", Type);
+            THROW_ERRNO(EINVAL);
+        }
+    }
+}
+
+void ProcessMessage(wsl::shared::SocketChannel& Channel, LX_MESSAGE_TYPE Type, const gsl::span<gsl::byte>& Buffer)
+{
+    try
+    {
+        HandleMessage<LSW_GET_DISK, LSW_MOUNT, LSW_EXEC, LSW_FORK, LSW_CONNECT, LSW_WAITPID, LSW_SIGNAL, LSW_TTY_RELAY>(Channel, Type, Buffer);
+    }
+    catch (...)
+    {
+        LOG_CAUGHT_EXCEPTION();
+
+        // TODO: error message
+    }
+}
+
+void ProcessMessages(wsl::shared::SocketChannel& Channel)
+{
+    while (Channel.Connected())
+    {
+        auto [Message, Range] = Channel.ReceiveMessageOrClosed<MESSAGE_HEADER>();
+        if (Message == nullptr || Message->MessageType == LxMessageLswShutdown)
+        {
+            break;
+        }
+
+        ProcessMessage(Channel, Message->MessageType, Range);
+    }
+
+    LOG_INFO("Process {} exiting", getpid());
+}
+
+int LswEntryPoint(int Argc, char* Argv[])
+{
+
+    //
+    // Mount devtmpfs.
+    //
+
+    int Result = UtilMount(nullptr, "/dev", "devtmpfs", 0, nullptr);
```

Comment:
We might want to think about if we want to mount some virtual filesystems after chrooting.